parallel_algorithms/P3732/numeric_ranges_algorithms.md

```diff
@@ -0,0 +1,1256 @@
+initial value or an identity for their custom operations.
+
+In the [Draft Fortran 2023 Standard](https://j3-fortran.org/doc/year/23/23-007r1.pdf), the `REDUCE` clause
+permits specification of an identity element.
+
+OpenMP lets users specify the identity value (via an _initializer-clause_ `initializer(`_initializer-expr_`)`), which is
+"used as the initializer for private copies of reduction list items" (see the relevant section of the [OpenMP 5.0
+specification](https://www.openmp.org/spec-html/5.0/openmpsu107.html)).
+
+Kokkos lets users define the identity value for custom reduction result types, by giving the reducer class an
+`init(value_type& value)` member function that sets `value` to the identity (see the [section on custom reducers
+in the Kokkos Programming Guide](https://kokkos.org/kokkos-core-wiki/ProgrammingGuide/Custom-Reductions-Custom-Reducers.html)).
+
+oneTBB asks users to specify the identity value as an argument to `parallel_reduce` function template (see the [relevant
+oneTBB specification page](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onetbb/source/algorithms/functions/parallel_reduce_func)).
+
+SYCL lets users specify the identity value by specializing `sycl::known_identity` class template for a custom reduction operation
+(see the [relevant section of the SYCL specification](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reduction)).
+
+The `std::linalg` linear algebra library in the Working Draft for C++26 says, "A value-initialized object of linear algebra
+value type shall act as the additive identity" ([linalg.reqs.val]{- .sref} 3).
+
+In Python's NumPy library, [`numpy.ufunc.reduce`](https://numpy.org/doc/stable/reference/generated/numpy.ufunc.reduce.html) takes optional
+initial values. If not provided and the binary operation (a "universal function" (ufunc), effectively an elementwise binary
+operation on a possibly multidimensional array) has an identity, then the initial values default to the identity. If the
+binary operation has no identity or the initial values are `None`, then this works like `reduce_first`.
+
+#### Conclusions
+
+Based on the above considerations, we conclude that there are good reasons to consider a mechanism for users to explicitly
+specify the identity element for parallel reduction. There are options of how that could be achieved, of which we list a few.
+
+- Add an optional extra parameter for the value of identity, defaulting to value initialization.
+- Change the meaning of the `init` parameter for parallel algorithms to represent identity instead of the initial value.
+- Provide a customization point similar to `sycl::known_identity` that also defaults to value initialization but can be
+  specialized for a given operation.
+- Similarly to `std::linalg`, require that for numeric parallel algorithms a value-initialized object shall act as the identity element.

```

Comment:
Our SG1 talk spoke of other ways, like structs with specializations.  That discussion can go in the next version of the paper, though.